README.md

<!-- README.md is generated from README.Rmd. Please edit that file and knit again -->
<!-- badges: start -->

[![Lifecycle:
stable](https://img.shields.io/badge/lifecycle-stable-brightgreen.svg)](https://lifecycle.r-lib.org/articles/stages.html#stable)
[![CRAN_Status_Badge](http://www.r-pkg.org/badges/version/heplots)](http://cran.r-project.org/package=heplots)
[![](http://cranlogs.r-pkg.org/badges/grand-total/heplots)](https://cran.r-project.org/package=heplots)
[![License](https://img.shields.io/badge/license-GPL%20(%3E=%202)-brightgreen.svg?style=flat)](https://www.gnu.org/licenses/gpl-2.0.html)
[![DOI](https://zenodo.org/badge/13908453.svg)](https://zenodo.org/badge/latestdoi/13908453)

<!-- badges: end -->

# heplots <img src="man/figures/logo.png" height="200" style="float:right; height:200px;"/>

## **Visualizing Hypothesis Tests in Multivariate Linear Models**

Version 1.5.0

## Description

The `heplots` package provides functions for visualizing hypothesis
tests in multivariate linear models (MANOVA, multivariate multiple
regression, MANCOVA, and repeated measures designs).

HE plots represent sums-of-squares-and-products matrices for linear
hypotheses (**H**) and for error (**E**) using ellipses (in two
dimensions), ellipsoids (in three dimensions), or by line segments in
one dimension. For the theory and applications, see:

- Fox, Friendly and Monette (2009) for a brief introduction,
- Friendly (2010) for the application of these ideas to repeated measure
  designs,
- Friendly, Monette and Fox (2013) for a general discussion of the role
  of elliptical geometry in statistical understanding, and
- Friendly & Sigal (2017) for an applied R tutorial.

Other topics now addressed here include:

- robust MLMs, using iteratively re-weighted least squared to
  down-weight observations with large multivariate residuals,
  `robmlm()`.
- `Mahalanobis()` calculates classical and robust Mahalanobis squared
  distances using MCD and MVE estimators of center and covariance.
- visualizing tests for equality of covariance matrices in MLMs (Box’s M
  test), `boxM()` and `plot.boxM()`.
- $\chi^2$ Q-Q plots for MLMs (`cqplot()`) to detect outliers and assess
  multivariate normality of residuals.
- bivariate coefficient plots showing elliptical confidence regions
  (`coefplot()`).

In this respect, the `heplots` package now aims to provide a wide range
of tools for analyzing and visualizing multivariate response linear
models, together with other packages:

<a href="http://github.com/friendly/candisc"><img src='https://raw.githubusercontent.com/friendly/candisc/master/candisc-logo.png' height='80'  style="float:right; height:80px;"></a>

- The related [`candisc`](http://github.com/friendly/candisc) package
  provides HE plots in **canonical discriminant** space, the space of
  linear combinations of the responses that show the maximum possible
  effects and for canonical correlation in multivariate regression
  designs.

<a href="https://friendly.github.io/mvinfluence/"><img src='https://raw.githubusercontent.com/friendly/mvinfluence/master/man/figures/logo.png' height='80'  style="float:right; height:80px;"></a>

- Another package,
  [`mvinfluence`](https://friendly.github.io/mvinfluence/), provides
  diagnostic measures and plots for **influential observations** in MLM
  designs.

Several tutorial vignettes are also included. See
`vignette(package="heplots")`.

## Installation

|                     |                                               |
|---------------------|-----------------------------------------------|
| CRAN version        | `install.packages("heplots")`                 |
| Development version | `remotes::install_github("friendly/heplots")` |

<!-- This installs the package from the source and creates the package vignettes,  -->
<!-- so you will need to have R Tools installed on your system.  [R Tools for Windows](https://cran.r-project.org/bin/windows/Rtools/) -->
<!-- takes you to the download page for Windows.  [R Tools for Mac OS X](https://cran.r-project.org/bin/macosx/tools/) -->
<!-- has the required programs for Mac OS X. -->

## HE plot functions

The graphical functions contained here all display multivariate model
effects in variable (**data**) space, for one or more response variables
(or contrasts among response variables in repeated measures designs).

- `heplot()` constructs two-dimensional HE plots for model terms and
  linear hypotheses for pairs of response variables in multivariate
  linear models.

- `heplot3d()` constructs analogous 3D plots for triples of response
  variables.

- The `pairs` method, `pairs.mlm()` constructs a scatterplot matrix of
  pairwise HE plots.

- `heplot1d()` constructs 1-dimensional analogs of HE plots for model
  terms and linear hypotheses for single response variables.

## Other functions

- `glance.mlm()` extends `broom::glance.lm()` to multivariate response
  models, giving a one-line statistical summary for each response
  variable.

- `boxM()` Calculates Box’s *M* test for homogeneity of covariance
  matrices in a MANOVA design. A `plot` method displays a visual
  representation of the components of the test. Associated with this,
  `bartletTests()` and `levineTests()` give the univariate tests of
  homogeneity of variance for each response measure in a MLM.

- `covEllipses()` draw covariance (data) ellipses for one or more group,
  optionally including the ellipse for the pooled within-group
  covariance.

### Repeated measure designs

For repeated measure designs, between-subject effects and within-subject
effects must be plotted separately, because the error terms (**E**
matrices) differ. For terms involving within-subject effects, these
functions carry out a linear transformation of the matrix **Y** of
responses to a matrix **Y M**, where **M** is the model matrix for a
term in the intra-subject design and produce plots of the H and E
matrices in this transformed space. The vignette `"repeated"` describes
these graphical methods for repeated measures designs. (At present, this
vignette is only available at [HE plots for repeated measures
designs](http://www.jstatsoft.org/v37/i04/paper).)

## Datasets

The package also provides a large collection of data sets illustrating a
variety of multivariate linear models of the types listed above,
together with graphical displays. The table below classifies these with
method tags. Their names are linked to the documentation on the
`pkgdown` website, \[<http://friendly.github.io/heplots>\].

| dataset                                                                               | rows | cols | title                                                       | tags             |
|:--------------------------------------------------------------------------------------|-----:|-----:|:------------------------------------------------------------|:-----------------|
| [AddHealth](http://friendly.github.io/heplots/reference/AddHealth.html)               | 4344 |    3 | Adolescent Health Data                                      | MANOVA ordered   |
| [Adopted](http://friendly.github.io/heplots/reference/Adopted.html)                   |   62 |    6 | Adopted Children                                            | MMRA repeated    |
| [Bees](http://friendly.github.io/heplots/reference/Bees.html)                         |  246 |    6 | Captive and maltreated bees                                 | MANOVA           |
| [Diabetes](http://friendly.github.io/heplots/reference/Diabetes.html)                 |  145 |    6 | Diabetes Dataset                                            | MANOVA           |
| [FootHead](http://friendly.github.io/heplots/reference/FootHead.html)                 |   90 |    7 | Head measurements of football players                       | MANOVA contrasts |
| [Headache](http://friendly.github.io/heplots/reference/Headache.html)                 |   98 |    6 | Treatment of Headache Sufferers for Sensitivity to Noise    | MANOVA repeated  |
| [Hernior](http://friendly.github.io/heplots/reference/Hernior.html)                   |   32 |    9 | Recovery from Elective Herniorrhaphy                        | MMRA candisc     |
| [Iwasaki_Big_Five](http://friendly.github.io/heplots/reference/Iwasaki_Big_Five.html) |  203 |    7 | Personality Traits of Cultural Groups                       | MANOVA           |
| [MockJury](http://friendly.github.io/heplots/reference/MockJury.html)                 |  114 |   17 | Effects Of Physical Attractiveness Upon Mock Jury Decisions | MANOVA candisc   |
| [NLSY](http://friendly.github.io/heplots/reference/NLSY.html)                         |  243 |    6 | National Longitudinal Survey of Youth Data                  | MMRA             |
| [NeuroCog](http://friendly.github.io/heplots/reference/NeuroCog.html)                 |  242 |   10 | Neurocognitive Measures in Psychiatric Groups               | MANOVA candisc   |
| [Oslo](http://friendly.github.io/heplots/reference/Oslo.html)                         |  332 |   14 | Oslo Transect Subset Data                                   | MANOVA candisc   |
| [Parenting](http://friendly.github.io/heplots/reference/Parenting.html)               |   60 |    4 | Father Parenting Competence                                 | MANOVA contrasts |
| [Plastic](http://friendly.github.io/heplots/reference/Plastic.html)                   |   20 |    5 | Plastic Film Data                                           | MANOVA           |
| [Pottery2](http://friendly.github.io/heplots/reference/Pottery2.html)                 |   48 |   12 | Chemical Analysis of Romano-British Pottery                 | MANOVA candisc   |
| [Probe1](http://friendly.github.io/heplots/reference/Probe1.html)                     |   11 |    5 | Response Speed in a Probe Experiment                        | MANOVA repeated  |
| [Probe2](http://friendly.github.io/heplots/reference/Probe2.html)                     |   20 |    6 | Response Speed in a Probe Experiment                        | MANOVA repeated  |
| [RatWeight](http://friendly.github.io/heplots/reference/RatWeight.html)               |   27 |    6 | Weight Gain in Rats Exposed to Thiouracil and Thyroxin      | MANOVA repeated  |
| [ReactTime](http://friendly.github.io/heplots/reference/ReactTime.html)               |   10 |    6 | Reaction Time Data                                          | repeated         |
| [Rohwer](http://friendly.github.io/heplots/reference/Rohwer.html)                     |   69 |   10 | Rohwer Data Set                                             | MMRA MANCOVA     |
| [RootStock](http://friendly.github.io/heplots/reference/RootStock.html)               |   48 |    5 | Growth of Apple Trees from Different Root Stocks            | MANOVA contrasts |
| [Sake](http://friendly.github.io/heplots/reference/Sake.html)                         |   30 |   10 | Taste Ratings of Japanese Rice Wine (Sake)                  | MMRA             |
| [Skulls](http://friendly.github.io/heplots/reference/Skulls.html)                     |  150 |    5 | Egyptian Skulls                                             | MANOVA contrasts |
| [SocGrades](http://friendly.github.io/heplots/reference/SocGrades.html)               |   40 |   10 | Grades in a Sociology Course                                | MANOVA candisc   |
| [SocialCog](http://friendly.github.io/heplots/reference/SocialCog.html)               |  139 |    5 | Social Cognitive Measures in Psychiatric Groups             | MANOVA candisc   |
| [TIPI](http://friendly.github.io/heplots/reference/TIPI.html)                         | 1799 |   16 | Data on the Ten Item Personality Inventory                  | MANOVA candisc   |
| [VocabGrowth](http://friendly.github.io/heplots/reference/VocabGrowth.html)           |   64 |    4 | Vocabulary growth data                                      | repeated         |
| [WeightLoss](http://friendly.github.io/heplots/reference/WeightLoss.html)             |   34 |    7 | Weight Loss Data                                            | repeated         |
| [mathscore](http://friendly.github.io/heplots/reference/mathscore.html)               |   12 |    3 | Math scores for basic math and word problems                | MANOVA           |
| [schooldata](http://friendly.github.io/heplots/reference/schooldata.html)             |   70 |    8 | School Data                                                 | MMRA robust      |

## Examples

This example illustrates HE plots using the classic `iris` data set. How
do the means of the flower variables differ by `Species`? This dataset
was the imputus for R. A. Fisher (1936) to propose a method of
discriminant analysis using data collected by Edgar Anderson (1928).
Though some may rightly deprecate Fisher for being a supporter of
eugenics, Anderson’s `iris` dataset should not be blamed.

A basic HE plot shows the **H** and **E** ellipses for the first two
response variables (here: `Sepal.Length` and `Sepal.Width`). The
multivariate test is significant (by Roy’s test) *iff* the **H** ellipse
projects *anywhere* outside the **E** ellipse.

The positions of the group means show how they differ on the two
response variables shown, and provide an interpretation of the
orientation of the **H** ellipse: it is long in the directions of
differences among the means.

``` r
iris.mod <- lm(cbind(Sepal.Length, Sepal.Width, Petal.Length, Petal.Width) ~ 
                 Species, data=iris)
heplot(iris.mod)
```

<img src="man/figures/README-iris1-1.png" width="70%" />

Contrasts or other linear hypotheses can be shown as well, and the
ellipses look better if they are filled. We create contrasts to test the
differences between `versacolor` and `virginca` and also between
`setosa` and the average of the other two. Each 1 df contrast plots as
degenerate 1D ellipse– a line.

Because these contrasts are orthogonal, they add to the total 2 df
effect of `Species`. Note how the first contrast, labeled `V:V`,
distinguishes the means of *versicolor* from *virginica*; the second
contrast, `S:VV` distinguishes `setosa` from the other two.

``` r
par(mar=c(4,4,1,1)+.1)
contrasts(iris$Species)<-matrix(c(0, -1, 1, 
                                  2, -1, -1), nrow=3, ncol=2)
contrasts(iris$Species)
#>            [,1] [,2]
#> setosa        0    2
#> versicolor   -1   -1
#> virginica     1   -1
iris.mod <- lm(cbind(Sepal.Length, Sepal.Width, Petal.Length, Petal.Width) ~ 
                 Species, data=iris)

hyp <- list("V:V"="Species1","S:VV"="Species2")
heplot(iris.mod, hypotheses=hyp, 
       fill=TRUE, fill.alpha=0.1)
```

<img src="man/figures/README-iris2-1.png" width="70%" />

All pairwise HE plots are produced using the `pairs` method for MLM
objects.

``` r
pairs(iris.mod, hypotheses=hyp, hyp.labels=FALSE,
      fill=TRUE, fill.alpha=0.1)
```

<img src="man/figures/README-iris3-1.png" width="100%" />

## References

Anderson, E. (1928). The Problem of Species in the Northern Blue Flags,
Iris versicolor L. and Iris virginica L. *Annals of the Missouri
Botanical Garden*, **13**, 241–313.

Fisher, R. A. (1936). The Use of Multiple Measurements in Taxonomic
Problems. *Annals of Eugenics*, **8**, 379–388.

Fox, J., Friendly, M. & Monette, G. (2009). [Visualizing hypothesis
tests in multivariate linear models: The heplots package for
R](http://datavis.ca/palers/FoxFriendlyMonette-2009.pdf) *Computational
Statistics*, **24**, 233-246.

Friendly, M. (2006). [Data Ellipses, HE Plots and Reduced-Rank Displays
for Multivariate Linear Models: SAS Software and
Examples.](https://www.jstatsoft.org/article/view/v017i06) *Journal of
Statistical Software*, **17**, 1-42.

Friendly, M. (2010). [HE plots for repeated measures
designs](http://www.jstatsoft.org/v37/i04/paper). *Journal of
Statistical Software*, **37**, 1–37.

Friendly, M.; Monette, G. & Fox, J. (2013). [Elliptical Insights:
Understanding Statistical Methods Through Elliptical
Geometry](http://datavis.ca/palers/ellipses-STS402.pdf) *Statistical
Science*, **28**, 1-39.

Friendly, M. & Sigal, M. (2017). [Graphical Methods for Multivariate
Linear Models in Psychological Research: An R
Tutorial.](https://doi.org/10.20982/tqmp.13.1.p020) *The Quantitative
Methods for Psychology*, **13**, 20-45.


NEWS
Version 1.3-6 (2019-03-26)

* add mathscore data

Version 1.3-6 (2019-03-26)

* fix bug in etasq.lm per #1 [thx: Phil Chalmers]
* bump pkg version
* remove Suggests: FRB

Version 1.3-5 (2018-04-02)

* fix references to car datatsets -> carData
* fix error resulting from car 3.0 changes
* bump pkg version

Version 1.3-4 (2016-11-7)

o Added Diabetes data, a nice example of heterogeneity of covariances

Version 1.3-3 (2016-10-21)

o Added Parenting data, a simple one-way MANOVA design (used in TQMP paper)
o Prepare for CRAN release

Version 1.3-1 (2016-06-10)

o added `plot.robmlm()`, a plot method for "robmlm" objects.  Currently, it plots the weights given
  to observations in the last iteration as an index plot.
o Re-built vignettes

Version 1.3-0 (2016-06-03)

o In cqplot, pch, col, and cex can now be vectors
o Bump version, prepare for release

Version 1.2-1 (2016-05-19)

o in coefplot.mlm(), now pass `label.pos` to label.ellipse()
o added Mahalanobis() for classical and robust squared distances; handles 
  missing data gracefully and provides a confidence envelope
o added SocialCog data [Thx: Leah Hartman]
o added cqplot() of Mahalanobis distances as a plot method for an mlm and for 
  multivariate data

Version 1.2-0 (2016-04-27)

o covEllipses() extended to more than two variables, giving a scatterplot matrix plot
o plot.boxM() now can plot other measures of the eigenvalues of the covariance matrices,
  useful for understanding the properties of the test.
o added bartlettTests() for a collection of univariate Bartlett tests 
o added leveneTests() for a collection of univariate Levene tests 
o added NeuroCog data, a simple one-way MANOVA [Thx: Leah Hartman]
o label.ellipse() now uses a much more flexible `label.pos` argument for positioning the
  text labels used in heplot() and friends.

Version 1.1-0 (2016-03-03)

o Added boXM from biotools
o Fleshed out boxM() with formula, lm, plot and summary methods
o added plot.boxM() method
o Added covEllipses() methods for plotting (possibly robust) covariance ellipses
o Added colDevs() as a utility and for use in other multivariate tests.
o Added data(AddHealth), a simple example of a one-way MANOVA with nice views [thx: Russell Warne]
o summary.boxM() now returns the calculated statistics and can be used to construct other kinds of plots.
o plot.boxM() gets a gplabel argument to provide a label for the group factor
o pairs.mlm() gets a var.cex argument to control size of variable labels
o Added logdetCI() for confidence intervals on log determinants of covariance matrices (experimental)
o Now use logdetCI() in plot.boxM()


Version 1.0-16 (2015-07-11)
o use importsFrom for base packages

Version 1.0-15 (2015-04-15)
o Fix require -> requireNamespace

Version 1.0-14 (2015-02-15)
o Updated broken urls in documentation
o Removed vcov.mlm from exports to conform to CRAN policies
o Now use rgl:: for rgl functions in *3d.R
o Removed robmlm.default, robmlm.formula from exports
o Imports: rgl


Version 1.0-12 (2014-05-15)
o Removed Authors: to satisfy CRAN nits
o Fixed doc for Hernior.Rd
o Now importsFrom(MASS) rather than Depends:


Version 1.0-11 (2013-08-20)
o Cosmetic changes to pass CRAN checks
o For R 3.1.0, revised vcov.mlm() to avoid calling summary.mlm() directly
o fixed long lines in .Rd files for R 3.0.1 checks
o removed require() calls for MASS & car (Depends:)
o removed ::: calls to car and stats

Version 1.0-10 (2013-06-03)
o Added demo/HE-other.R, exploring other representations of HE plots

Version 1.0-9 (2013-04-19)
o Added Pottery2 data, extending car::Pottery
o Fixed heplot3d to minimize striated rendering of ellipsoids, and also modified default
  colors and fill.alpha

Version 1.0-8 (2013-02-22)
o In heplot.mlm(), added ... to label.ellipse to be able to pass cex=
o Added NLSY data with examples of testing linearHypotheses
o In heplot3d(), rgl.* functions replaced with *3d() equivalents
o Extended gsorth description and examples
 

Version 1.0-7 (2013-02-07)
o Made some heplot3d() examples \dontrun{} to reduce example execution time to satisfy CRAN nits

Version 1.0-6 (2013-01-14)
o Added label.ellipse() as a more general exported function.
o heplot() gets a label.pos= argument allowing more flexible labeling, with
  the label position to be specified (bottom, left, top, right or center)
o Added coefplot S3 generic and a coefplot.mlm method to plot ellipses for mlm coefficients.
  This generalizes car::confidenceEllipse to the mlm setting.

Version 1.0-4 (2012-12-29)
o Added interpPlot for plotting interpolations between two data sets, with data ellipses, etc.
  Main use for this is in animated demos of multivariate statistical concepts.
o interpPlot gains add=, points= and col= arguments for greater control of graphical parameters.

Version 1.0-2 (2012-12-05)
o Added schooldata and further example of comparison of mlm with robmlm()

Version 1.0-1 (2012-10-18)
o Added simple robmlm methods for robust mlms with print and summary methods and a vcov.mlm method.
  This allows heplot() and related methods to show robust HE plots.

Version 1.0-0 (2012-6-3)
o Added Olso data
o Corrected error in demo/OBK.R

Version 0.9-13 (2012-5-18)
o Extended HE-examples vignette with example based on SocGrades data

Version 0.9-12 (2012-04-03)
o Added Skulls data
o Extended HE-examples vignette with example based on Skulls data
o Changed the default colors for heplot() and heplot1d() to allow global setting
  of options("heplot.colors"); similarly, heplot3d() now allows a global option
  for colors, options("heplot3d.colors"), differing only in using pink for error
  ellipsoid.
o Extended Hernior.Rd examples to show 1D plots in canonical space; added a discussion of
  this example to HE-examples vignette.
o Extended SocGrades.Rd examples to show plots in canonical space.

Version 0.9-11 (2012-01-12)
o Added Probe1, Probe2 data sets-- repeated measures design
o Added Sake data

Version 0.9-10 (2011-09-12)

R/robmlm.R
### robust multivariate linear regression

## John Fox 2012-06-02
## revised: 2013-08-20 to avoid calling summary.mlm() directly in vcov.mlm()



#' Robust Fitting of Multivariate Linear Models
#' 
#' Fit a multivariate linear model by robust regression using a simple M
#' estimator.
#' 
#' These S3 methods are designed to provide a specification of a class of
#' robust methods which extend \code{mlm}s, and are therefore compatible with
#' other \code{mlm} extensions, including \code{\link[car]{Anova}} and
#' \code{\link{heplot}}.
#' 
#' Fitting is done by iterated re-weighted least squares (IWLS), using weights
#' based on the Mahalanobis squared distances of the current residuals from the
#' origin, and a scaling (covariance) matrix calculated by
#' \code{\link[MASS]{cov.trob}}. The design of these methods were loosely
#' modeled on \code{\link[MASS]{rlm}}.
#' 
#' An internal \code{vcov.mlm} function is an extension of the standard
#' \code{\link[stats]{vcov}} method providing for observation weights.
#' 
#' @aliases print.robmlm print.summary.robmlm robmlm robmlm.default
#' robmlm.formula summary.robmlm
#' @param formula a formula of the form \code{cbind(y1, y2, ...) ~ x1 + x2 +
#' ...}.
#' @param data a data frame from which variables specified in \code{formula}
#' are preferentially to be taken.
#' @param subset An index vector specifying the cases to be used in fitting.
#' @param weights a vector of prior weights for each case.
#' @param na.action A function to specify the action to be taken if \code{NA}s
#' are found.  The 'factory-fresh' default action in R is
#' \code{\link[stats]{na.omit}}, and can be changed by
#' \code{\link[base]{options}}\code{(na.action=)}.
#' @param model should the model frame be returned in the object?
#' @param contrasts optional contrast specifications; see
#' \code{\link[stats]{lm}} for details.
#' @param \dots other arguments, passed down. In particular relevant control
#' arguments can be passed to the to the \code{robmlm.default} method.
#' @param X for the default method, a model matrix, including the constant (if
#' present)
#' @param Y for the default method, a response matrix
#' @param w prior weights
#' @param P two-tail probability, to find cutoff quantile for chisq (tuning
#' constant); default is set for bisquare weight function
#' @param tune tuning constant (if given directly)
#' @param max.iter maximum number of iterations
#' @param psi robustness weight function; \code{\link[MASS]{psi.bisquare}} is
#' the default
#' @param tol convergence tolerance, maximum relative change in coefficients
#' @param initialize modeling function to find start values for coefficients,
#' equation-by-equation; if absent WLS (\code{lm.wfit}) is used
#' @param verbose show iteration history? (\code{TRUE} or \code{FALSE})
#' @param x a \code{robmlm} object
#' @param object a \code{robmlm} object
#' @return An object of class \code{"robmlm"} inheriting from \code{c("mlm",
#' "lm")}.
#' 
#' This means that the returned \code{"robmlm"} contains all the components of
#' \code{"mlm"} objects described for \code{\link[stats]{lm}}, plus the
#' following: %% If it is a LIST, use \item{weights }{final observation
#' weights} \item{iterations }{number of iterations} \item{converged }{logical:
#' did the IWLS process converge?} %% ...
#' 
#' The generic accessor functions \code{\link[stats]{coefficients}},
#' \code{\link[stats]{effects}}, \code{\link[stats]{fitted.values}} and
#' \code{\link[stats]{residuals}} extract various useful features of the value
#' returned by \code{robmlm}.
#' @author John Fox; packaged by Michael Friendly
#' @seealso
#' 
#' \code{\link[MASS]{rlm}}, \code{\link[MASS]{cov.trob}}
#' @references A. Marazzi (1993) \emph{Algorithms, Routines and S Functions for
#' Robust Statistics}.  Wadsworth & Brooks/Cole.
#' @keywords multivariate robust
#' @examples
#' 
#' ##############
#' # Skulls data
#' 
#' # make shorter labels for epochs and nicer variable labels in heplots
#' Skulls$epoch <- factor(Skulls$epoch, labels=sub("c","",levels(Skulls$epoch)))
#' # variable labels
#' vlab <- c("maxBreadth", "basibHeight", "basialLength", "nasalHeight")
#' 
#' # fit manova model, classically and robustly
#' sk.mod <- lm(cbind(mb, bh, bl, nh) ~ epoch, data=Skulls)
#' sk.rmod <- robmlm(cbind(mb, bh, bl, nh) ~ epoch, data=Skulls)
#' 
#' # standard mlm methods apply here
#' coefficients(sk.rmod)
#' 
#' # index plot of weights
#' plot(sk.rmod$weights, type="h", xlab="Case Index", ylab="Robust mlm weight", col="gray")
#' points(sk.rmod$weights, pch=16, col=Skulls$epoch)
#' axis(side=1, at=15+seq(0,120,30), labels=levels(Skulls$epoch), tick=FALSE, cex.axis=1)
#' 
#' # heplots to see effect of robmlm vs. mlm
#' heplot(sk.mod, hypotheses=list(Lin="epoch.L", Quad="epoch.Q"), 
#'     xlab=vlab[1], ylab=vlab[2], cex=1.25, lty=1)
#' heplot(sk.rmod, hypotheses=list(Lin="epoch.L", Quad="epoch.Q"), 
#'     add=TRUE, error.ellipse=TRUE, lwd=c(2,2), lty=c(2,2), 
#'     term.labels=FALSE, hyp.labels=FALSE, err.label="")
#' 
#' ##############
#' # Pottery data
#' 
#' pottery.mod <- lm(cbind(Al,Fe,Mg,Ca,Na)~Site, data=Pottery)
#' pottery.rmod <- robmlm(cbind(Al,Fe,Mg,Ca,Na)~Site, data=Pottery)
#' Anova(pottery.mod)
#' Anova(pottery.rmod)
#' 
#' # index plot of weights
#' plot(pottery.rmod$weights, type="h")
#' points(pottery.rmod$weights, pch=16, col=Pottery$Site)
#' 
#' # heplots to see effect of robmlm vs. mlm
#' heplot(pottery.mod, cex=1.3, lty=1)
#' heplot(pottery.rmod, add=TRUE, error.ellipse=TRUE, lwd=c(2,2), lty=c(2,2), 
#'     term.labels=FALSE, err.label="")
#' 
#' ###############
#' # Prestige data
#' 
#' # treat women and prestige as response variables for this example
#' prestige.mod <- lm(cbind(women, prestige) ~ income + education + type, data=Prestige)
#' prestige.rmod <- robmlm(cbind(women, prestige) ~ income + education + type, data=Prestige)
#' 
#' coef(prestige.mod)
#' coef(prestige.rmod)
#' # how much do coefficients change?
#' round(coef(prestige.mod) - coef(prestige.rmod),3)
#' 
#' # pretty plot of case weights
#' plot(prestige.rmod$weights, type="h", xlab="Case Index", ylab="Robust mlm weight", col="gray")
#' points(prestige.rmod$weights, pch=16, col=Prestige$type)
#' legend(0, 0.7, levels(Prestige$type), pch=16, col=palette()[1:3], bg="white")
#' 
#' heplot(prestige.mod, cex=1.4, lty=1)
#' heplot(prestige.rmod, add=TRUE, error.ellipse=TRUE, lwd=c(2,2), lty=c(2,2), 
#'     term.labels=FALSE, err.label="")
#' 
#' 
#' 
#' @export robmlm
robmlm <- function(X, ...){
  UseMethod("robmlm")
}

robmlm.default <- function(X, Y, w, P=2*pnorm(4.685, lower.tail=FALSE), 
                           tune, max.iter=100, psi=psi.bisquare, tol=1e-6, 
                           initialize, verbose=FALSE, ...){
  # args:
  #   X: model matrix, including constant (if present)
  #   Y: response matrix
  #   w: prior weights
  #   P: two-tail probability, to find cutoff quantile for chisq (tuning constant);
  #       default is set for bisquare weight function
  #   tune: tuning constant (if given directly)
  #   max.iter: iteration limit
  #   psi: robustness weight function; psi.bisquare (from MASS) is default
  #   tol: convergence tolerance, maximum relative change in coefficients
  #   initialize: modeling function to find start values for coefficients,
  #       equation-by-equation; if absent WLS is used
  #   verbose: show iteration history?
  #if (!require(MASS)) stop("MASS package missing")
  p <- ncol(Y)
  if (missing(w) || is.null(w)) w <- rep(1, nrow(Y))
  if (missing(tune)) tune <- qchisq(1 - P, df=p)
  if (missing(initialize)){
    fit.last <- lm.wfit(X, Y, w)
  }
  else{
    k <- ncol(X)
    n <- nrow(X)
    coef <- matrix(0, k, p)
    res <- matrix(0, n, p)
    for (j in 1:p){
      modj <- initialize(Y[, j] ~ X - 1, weights=w)
      coef[, j] <- coef(modj)
      res[, j] <- residuals(modj)
    }
    fit.last <- list(coefficients=coef, residuals=res)
  }
  B.last <- B.new <- fit.last$coefficients
  iter <- 0
  if (verbose){
    coefnames <- abbreviate(outer(rownames(B.new), colnames(B.new), function(x, y) paste(x, ":", y, sep="")), 10)
    b <- as.vector(B.new)
    names(b) <- coefnames
    cat("\n", iter, ":\n", sep="")
    print(b)
  }
  repeat {
    iter <- iter + 1
    if (iter > max.iter) break
    E <- fit.last$residuals
    S <- cov.trob(E, center=FALSE)$cov
    mahal <- mahalanobis(E, 0, S)
    wts <- psi(mahal, tune)
    fit.new <- lm.wfit(X, Y, w*wts)
    B.last <- B.new
    B.new <- fit.new$coefficients
    if (verbose){
      b <- as.vector(B.new)
      names(b) <- coefnames
      cat("\n", iter, ":\n", sep="")
      print(b)
    }
    if (max(abs((B.last - B.new)/(B.last + tol))) < tol) break
    fit.last <- fit.new
  }
  if (iter > max.iter) warning("maximum iterations exceeded")
  fit.new$iterations <- iter
  fit.new$weights <- wts
  fit.new$converged <- iter <= max.iter
  fit.new
}

robmlm.formula <- function(formula, data, subset, weights, na.action, model = TRUE,
                         contrasts = NULL, ...) {
  # ... passed to robmlm.default
  call <- match.call()  
  call[[1]] <- as.name("robmlm")
  mf <- match.call(expand.dots = FALSE)  
  args <- match(c("formula", "data", "subset", "weights", "na.action"),
                names(mf), 0)  
  mf <- mf[c(1, args)]
  mf$drop.unused.levels <- TRUE
  mf[[1]] <- as.name("model.frame")
  mf <- eval.parent(mf)  
  terms <- attr(mf, "terms")  
  Y <- model.response(mf, type="numeric")  
  X <- model.matrix(terms, mf, contrasts) 
  w <- as.vector(model.weights(mf))
  mod <- robmlm.default(X, Y, w, ...)
  mod$na.action <- attr(mf, "na.action")
  mod$contrasts <- attr(X, "contrasts")
  mod$xlevels <- .getXlevels(terms, mf)
  mod$call <- call
  mod$terms <- terms
  if (model)  mod$model <- mf
  class(mod) <- c("robmlm", "mlm", "lm")
  mod
}

print.robmlm <- function(x, ...){
  if (!x$converged) warning("failed to converge")
  NextMethod()
  cat("iterations = ", x$iterations)
  invisible(x)
}

summary.robmlm <- function(object, ...){
  res <- list()
  res[[1]] <- NextMethod()
  res$iterations <- object$iterations
  res$converged <- object$converged
  class(res) <- c("summary.robmlm", class(res[[1]]))
  res
}

print.summary.robmlm <- function(x, ...){
  if (!x$converged) warning("failed to converge")
  print(x[[1]])
  cat("iterations = ", x$iterations)
  invisible(x)
}

vcov.mlm <- function (object, ...) {
# override stats::vcov.mlm to allow weights
#   adapted from code provided by Michael Friendly
# For R 3.1.0, to avoid calling summary.mlm directly, change the object class
#   temporarily to c("mlm", "lm")
	SSD.mlm <- function (object, ...) {
		if (!is.null(object$weights)) { 
			SSD <- wcrossprod(residuals(object), w=object$weights)
			df <- sum(object$weights)
		}
		else {
			SSD <- crossprod(residuals(object))
			df <- object$df.residual
		}
		structure(list(SSD=SSD, call = object$call, 
						df = df), class = "SSD")
	}
	estVar.mlm <- function (object, ...) estVar(SSD(object))
	obj <- object
	class(obj) <- c("mlm", "lm")  # remove robmlm class temporarily
	so <- summary(obj)[[1L]]
	kronecker(estVar(object), so$cov.unscaled, make.dimnames = TRUE)
}
